File: db4on/core/src/core/com/db4o/YapServerThread.cs
namespace com.db4o
{
	internal sealed class YapServerThread : j4o.lang.Thread
	{
		private string i_clientName;

		private bool i_loggedin;

		private long i_lastClientMessage;

		private readonly com.db4o.YapFile i_mainStream;

		private com.db4o.Transaction i_mainTrans;

		private int i_pingAttempts = 0;

		private int i_nullMessages;

		private bool i_rollbackOnClose = true;

		private bool i_sendCloseMessage = true;

		private readonly com.db4o.YapServer i_server;

		private com.db4o.foundation.network.YapSocket i_socket;

		private com.db4o.YapFile i_substituteStream;

		private com.db4o.Transaction i_substituteTrans;

		private com.db4o.Config4Impl i_config;

		internal readonly int i_threadID;

		internal YapServerThread(com.db4o.YapServer aServer, com.db4o.YapFile aStream, com.db4o.foundation.network.YapSocket
			 aSocket, int aThreadID, bool loggedIn)
		{
			i_loggedin = loggedIn;
			i_lastClientMessage = j4o.lang.JavaSystem.CurrentTimeMillis();
			i_server = aServer;
			i_config = (com.db4o.Config4Impl)i_server.Configure();
			i_mainStream = aStream;
			i_threadID = aThreadID;
			SetName("db4o message server " + aThreadID);
			i_mainTrans = new com.db4o.Transaction(aStream, aStream.GetSystemTransaction());
			try
			{
				i_socket = aSocket;
				i_socket.SetSoTimeout(((com.db4o.Config4Impl)aServer.Configure()).TimeoutServerSocket
					());
			}
			catch (System.Exception e)
			{
				i_socket.Close();
				throw (e);
			}
		}

		public void Close()
		{
			CloseSubstituteStream();
			try
			{
				if (i_sendCloseMessage)
				{
					com.db4o.Msg.CLOSE.Write(i_mainStream, i_socket);
				}
			}
			catch (System.Exception e)
			{
			}
			if (i_mainStream != null && i_mainTrans != null)
			{
				i_mainTrans.Close(i_rollbackOnClose);
			}
			try
			{
				i_socket.Close();
			}
			catch (System.Exception e)
			{
			}
			i_socket = null;
			try
			{
				i_server.RemoveThread(this);
			}
			catch (System.Exception e)
			{
			}
		}

		private void CloseSubstituteStream()
		{
			if (i_substituteStream != null)
			{
				if (i_substituteTrans != null)
				{
					i_substituteTrans.Close(i_rollbackOnClose);
					i_substituteTrans = null;
				}
				try
				{
					i_substituteStream.Close();
				}
				catch (System.Exception e)
				{
				}
				i_substituteStream = null;
			}
		}

		private com.db4o.YapFile GetStream()
		{
			if (i_substituteStream != null)
			{
				return i_substituteStream;
			}
			return i_mainStream;
		}

		internal com.db4o.Transaction GetTransaction()
		{
			if (i_substituteTrans != null)
			{
				return i_substituteTrans;
			}
			return i_mainTrans;
		}

		public override void Run()
		{
			while (i_socket != null)
			{
				try
				{
					if (!MessageProcessor())
					{
						break;
					}
				}
				catch (System.IO.IOException e)
				{
					break;
				}
				catch (System.Exception e)
				{
					if (i_mainStream == null || i_mainStream.IsClosed())
					{
						break;
					}
				}
				if (i_nullMessages > 20 || PingClientTimeoutReached())
				{
					if (i_pingAttempts > 5)
					{
						GetStream().LogMsg(33, i_clientName);
						break;
					}
					if (null == i_socket)
					{
						break;
					}
					com.db4o.Msg.PING.Write(GetStream(), i_socket);
					i_pingAttempts++;
				}
			}
			Close();
		}

		private bool PingClientTimeoutReached()
		{
			return (j4o.lang.JavaSystem.CurrentTimeMillis() - i_lastClientMessage > i_config.
				TimeoutPingClients());
		}

		private bool MessageProcessor()
		{
			com.db4o.Msg message = com.db4o.Msg.ReadMessage(GetTransaction(), i_socket);
			if (message == null)
			{
				i_nullMessages++;
				return true;
			}
			i_lastClientMessage = j4o.lang.JavaSystem.CurrentTimeMillis();
			i_nullMessages = 0;
			i_pingAttempts = 0;
			if (!i_loggedin)
			{
				if (com.db4o.Msg.LOGIN.Equals(message))
				{
					string userName = ((com.db4o.MsgD)message).ReadString();
					string password = ((com.db4o.MsgD)message).ReadString();
					com.db4o.User user = new com.db4o.User();
					user.name = userName;
					i_mainStream.ShowInternalClasses(true);
					com.db4o.User found = (com.db4o.User)i_mainStream.Get(user).Next();
					i_mainStream.ShowInternalClasses(false);
					if (found != null)
					{
						if (found.password.Equals(password))
						{
							i_clientName = userName;
							i_mainStream.LogMsg(32, i_clientName);
							com.db4o.Msg.LOGIN_OK.GetWriterForInt(GetTransaction(), i_mainStream.BlockSize())
								.Write(i_mainStream, i_socket);
							i_loggedin = true;
							SetName("db4o server socket for client " + i_clientName);
						}
						else
						{
							com.db4o.Msg.FAILED.Write(i_mainStream, i_socket);
							return false;
						}
					}
					else
					{
						com.db4o.Msg.FAILED.Write(i_mainStream, i_socket);
						return false;
					}
				}
				return true;
			}
			if (message.ProcessMessageAtServer(i_socket))
			{
				return true;
			}
			if (com.db4o.Msg.PING.Equals(message))
			{
				com.db4o.Msg.OK.Write(GetStream(), i_socket);
				return true;
			}
			if (com.db4o.Msg.CLOSE.Equals(message))
			{
				com.db4o.Msg.CLOSE.Write(GetStream(), i_socket);
				GetTransaction().Commit();
				i_sendCloseMessage = false;
				GetStream().LogMsg(34, i_clientName);
				return false;
			}
			if (com.db4o.Msg.IDENTITY.Equals(message))
			{
				RespondInt((int)GetStream().GetID(GetStream().BootRecord().i_db));
				return true;
			}
			if (com.db4o.Msg.CURRENT_VERSION.Equals(message))
			{
				com.db4o.YapStream stream = GetStream();
				long ver = 0;
				lock (stream)
				{
					ver = GetStream().BootRecord().CurrentVersion();
				}
				com.db4o.Msg.ID_LIST.GetWriterForLong(GetTransaction(), ver).Write(GetStream(), i_socket
					);
				return true;
			}
			if (com.db4o.Msg.RAISE_VERSION.Equals(message))
			{
				long minimumVersion = ((com.db4o.MsgD)message).ReadLong();
				com.db4o.YapStream stream = GetStream();
				lock (stream)
				{
					stream.RaiseVersion(minimumVersion);
				}
				return true;
			}
			if (com.db4o.Msg.GET_THREAD_ID.Equals(message))
			{
				RespondInt(i_threadID);
				return true;
			}
			if (com.db4o.Msg.SWITCH_TO_FILE.Equals(message))
			{
				SwitchToFile(message);
				return true;
			}
			if (com.db4o.Msg.SWITCH_TO_MAIN_FILE.Equals(message))
			{
				SwitchToMainFile();
				return true;
			}
			if (com.db4o.Msg.USE_TRANSACTION.Equals(message))
			{
				UseTransaction(message);
				return true;
			}
			return true;
		}

		private void SwitchToFile(com.db4o.Msg message)
		{
			lock (i_mainStream.i_lock)
			{
				string fileName = ((com.db4o.MsgD)message).ReadString();
				try
				{
					CloseSubstituteStream();
					i_substituteStream = (com.db4o.YapFile)com.db4o.Db4o.OpenFile(fileName);
					i_substituteTrans = new com.db4o.Transaction(i_substituteStream, i_substituteStream
						.GetSystemTransaction());
					i_substituteStream.i_config.SetMessageRecipient(i_mainStream.i_config.MessageRecipient
						());
					com.db4o.Msg.OK.Write(GetStream(), i_socket);
				}
				catch (System.Exception e)
				{
					CloseSubstituteStream();
					com.db4o.Msg.ERROR.Write(GetStream(), i_socket);
				}
			}
		}

		private void SwitchToMainFile()
		{
			lock (i_mainStream.i_lock)
			{
				CloseSubstituteStream();
				com.db4o.Msg.OK.Write(GetStream(), i_socket);
			}
		}

		private void UseTransaction(com.db4o.Msg message)
		{
			int threadID = ((com.db4o.MsgD)message).ReadInt();
			com.db4o.YapServerThread transactionThread = i_server.FindThread(threadID);
			if (transactionThread != null)
			{
				com.db4o.Transaction transToUse = transactionThread.GetTransaction();
				if (i_substituteTrans != null)
				{
					i_substituteTrans = transToUse;
				}
				else
				{
					i_mainTrans = transToUse;
				}
				i_rollbackOnClose = false;
			}
		}

		private void RespondInt(int response)
		{
			com.db4o.Msg.ID_LIST.GetWriterForInt(GetTransaction(), response).Write(GetStream(
				), i_socket);
		}
	}
}

File: db4oj/core/src/com/db4o/YapServerThread.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o;

import java.io.IOException;

import com.db4o.foundation.network.*;

final class YapServerThread extends Thread {

    private String i_clientName;

    private boolean i_loggedin;
    private long i_lastClientMessage;
    private final YapFile i_mainStream;

    private Transaction i_mainTrans;
    private int i_pingAttempts = 0;
    private int i_nullMessages;
    private boolean i_rollbackOnClose = true;
    private boolean i_sendCloseMessage = true;

    private final YapServer i_server;

    private YapSocket i_socket;
    private YapFile i_substituteStream;
    private Transaction i_substituteTrans;
    
    private Config4Impl i_config;

    final int i_threadID;

    YapServerThread(
        YapServer aServer,
        YapFile aStream,
        YapSocket aSocket,
        int aThreadID,
        boolean loggedIn)
        throws Exception {
        
        	
        i_loggedin = loggedIn;
         
        i_lastClientMessage = System.currentTimeMillis(); // don't start pinging from the start
        i_server = aServer;
		i_config = (Config4Impl)i_server.configure();
        i_mainStream = aStream;
        i_threadID = aThreadID;
        setName("db4o message server " + aThreadID);
        i_mainTrans = new Transaction(aStream, aStream.getSystemTransaction());
        try {
            i_socket = aSocket;
            i_socket.setSoTimeout(((Config4Impl)aServer.configure()).timeoutServerSocket());

            // TODO: Experiment with packetsize and noDelay
            // i_socket.setSendBufferSize(100);
            // i_socket.setTcpNoDelay(true);

        } catch (Exception e) {
            i_socket.close();
            throw (e);
        }
    }

    public void close() {
        closeSubstituteStream();
        try {
            if (i_sendCloseMessage) {
                Msg.CLOSE.write(i_mainStream, i_socket);
            }
        } catch (Exception e) {
            if (Debug.atHome) {
                e.printStackTrace();
            }

        }
        if (i_mainStream != null && i_mainTrans != null) {
            i_mainTrans.close(i_rollbackOnClose);
        }
        try {
            i_socket.close();
        } catch (Exception e) {
            if (Debug.atHome) {
                e.printStackTrace();
            }
        }
        i_socket = null;
        try {
            i_server.removeThread(this);
        } catch (Exception e) {
            if (Debug.atHome) {
                e.printStackTrace();
            }
        }
    }

    private void closeSubstituteStream() {
        if (i_substituteStream != null) {
            if (i_substituteTrans != null) {
                i_substituteTrans.close(i_rollbackOnClose);
                i_substituteTrans = null;
            }
            try {
                i_substituteStream.close();

            } catch (Exception e) {
                if (Debug.atHome) {
                    e.printStackTrace();
                }
            }
            i_substituteStream = null;
        }
    }

    private final YapFile getStream() {
        if (i_substituteStream != null) {
            return i_substituteStream;
        }
        return i_mainStream;
    }

    Transaction getTransaction() {
        if (i_substituteTrans != null) {
            return i_substituteTrans;
        }
        return i_mainTrans;
    }

    public void run() {
        while (i_socket != null) {
            try {
                if(! messageProcessor()){
                    break;
                }
            } catch (Exception e) {
                if (i_mainStream == null || i_mainStream.isClosed()) {
                    break;
                }
                if(! i_socket.isConnected()){
                    break;
                }
                if (Deploy.debug) {
                    e.printStackTrace();
                }
                i_nullMessages++;
            }
            
            // TODO: Optimize - this doesn't need to be in the loop of executing statements

            if (i_nullMessages > 20 || pingClientTimeoutReached()) {
                if (i_pingAttempts > 5) {
                    // 
                    getStream().logMsg(33, i_clientName);
                    break;
                }
                if (null == i_socket) break;
                Msg.PING.write(getStream(), i_socket);
                i_pingAttempts++;
            }
        }
        close();
    }

	private boolean pingClientTimeoutReached() {
		return (System.currentTimeMillis() - i_lastClientMessage > i_config.timeoutPingClients());
	}
    
    private boolean messageProcessor() throws IOException{
        
        Msg message = Msg.readMessage(getTransaction(), i_socket);
        if(message == null){
            i_nullMessages ++;
            return true;
        }
        
        i_lastClientMessage = System.currentTimeMillis();
        i_nullMessages = 0;
        i_pingAttempts = 0;
        if (! i_loggedin) {
            if (Msg.LOGIN.equals(message)) {
                String userName = ((MsgD) message).readString();
                String password = ((MsgD) message).readString();
                User user = new User();
                user.name = userName;
                i_mainStream.showInternalClasses(true);
                User found = (User) i_mainStream.get(user).next();
                i_mainStream.showInternalClasses(false);
                if (found != null) {
                    if (found.password.equals(password)) {
                        i_clientName = userName;
                        i_mainStream.logMsg(32, i_clientName);
                        int blockSize = i_mainStream.blockSize();
                        int encrypt = i_mainStream.i_handlers.i_encrypt ? 1 : 0;
                        Msg.LOGIN_OK.getWriterForInts(getTransaction(), new int[] {blockSize, encrypt}).write(i_mainStream, i_socket);
                        i_loggedin= true;
                        setName("db4o server socket for client " + i_clientName);
                    } else {
                        Msg.FAILED.write(i_mainStream, i_socket);
                        return false;
                    }
                } else {
                    Msg.FAILED.write(i_mainStream, i_socket);
                    return false;
                }
            }
            return true;
        }
        
        if (message.processMessageAtServer(i_socket)) {
            return true;
        }
        
        if (Msg.PING.equals(message)) {
            Msg.OK.write(getStream(), i_socket);
            return true;
        }
        if (Msg.CLOSE.equals(message)) {
            Msg.CLOSE.write(getStream(), i_socket);
            getTransaction().commit();
            i_sendCloseMessage = false;
            getStream().logMsg(34, i_clientName);
            return false;
        }
        
        if (Msg.IDENTITY.equals(message)) {
            respondInt((int)getStream().getID(getStream().bootRecord().i_db));
            return true;
        }
        
        if (Msg.CURRENT_VERSION.equals(message)){
            long ver = 0;
            synchronized(getStream()){
                ver = getStream().currentVersion();
            }
            Msg.ID_LIST.getWriterForLong(getTransaction(), ver).write(getStream(), i_socket);
            return true;
        }
        
        if (Msg.RAISE_VERSION.equals(message)) {
            long minimumVersion = ((MsgD)message).readLong();
            YapStream stream = getStream();
            synchronized(stream){
                stream.raiseVersion(minimumVersion);
            }
            return true;
        } 
        
        if (Msg.GET_THREAD_ID.equals(message)) {
            respondInt(i_threadID);
            return true;
        }
        
        if (Msg.SWITCH_TO_FILE.equals(message)) {
            switchToFile(message);
            return true;
        }
        
        if (Msg.SWITCH_TO_MAIN_FILE.equals(message)) {
            switchToMainFile();
            return true;
        }
        
        if (Msg.USE_TRANSACTION.equals(message)) {
            useTransaction(message);
            return true;
        }
        
        return true;
    }
    
    

    private void switchToFile(Msg message) {
        synchronized (i_mainStream.i_lock) {
            String fileName = ((MsgD) message).readString();
            try {
                closeSubstituteStream();
                i_substituteStream = (YapFile) Db4o.openFile(fileName);
                i_substituteTrans =
                    new Transaction(i_substituteStream, i_substituteStream.getSystemTransaction());
                i_substituteStream.i_config.setMessageRecipient(i_mainStream.i_config.messageRecipient());
                Msg.OK.write(getStream(), i_socket);
            } catch (Exception e) {
                if (Debug.atHome) {
                    System.out.println("Msg.SWITCH_TO_FILE failed.");
                    e.printStackTrace();
                }
                closeSubstituteStream();
                Msg.ERROR.write(getStream(), i_socket);
            }
        }
    }

    private void switchToMainFile() {
        synchronized (i_mainStream.i_lock) {
            closeSubstituteStream();
            Msg.OK.write(getStream(), i_socket);
        }
    }

    private void useTransaction(Msg message) {
        int threadID = ((MsgD) message).readInt();
        YapServerThread transactionThread = i_server.findThread(threadID);
        if (transactionThread != null) {
            Transaction transToUse = transactionThread.getTransaction();
            if (i_substituteTrans != null) {
                i_substituteTrans = transToUse;
            } else {
                i_mainTrans = transToUse;
            }
            i_rollbackOnClose = false;
        }
    }
    
    private void respondInt(int response){
        Msg.ID_LIST.getWriterForInt(getTransaction(), response).write(getStream(),i_socket);
    }
    
    
}